File: TaskList.Backend/CombGuid.cs
using System.Data.SqlTypes;
using System.Security.Cryptography;

namespace TaskList.Backend;

public static class CombGuid
{
    /// <summary>
    /// Byte order SQL Server uses when comparing <see cref="Guid"/> values.
    /// </summary>
    private static readonly int[] SqlOrder = { 3, 2, 1, 0, 5, 4, 7, 6, 8, 9, 10, 11, 12, 13, 14, 15 };

    private static readonly object Lock = new();

    // Per-process counter to break ties within the same millisecond
    private static long _lastMilliseconds;
    private static uint _sequence;

    /// <summary>
    /// Generates a COMB GUID (Combined GUID/Timestamp) that maintains chronological order across systems while
    /// remaining globally unique.
    /// </summary>
    /// <remarks>
    /// <para>
    /// THIS FUNCTION IS DUPLICATED IN TetraTech.CNU.Database.ClrFunctions
    /// </para>
    /// <para>
    /// This function creates a 128-bit <see cref="Guid"/> where the bytes that SQL Server compares first (based
    /// on its internal byte ordering) contain a UTC timestamp in milliseconds since the Unix epoch. The remaining
    /// bytes are random, ensuring global uniqueness.
    /// </para>
    /// <para>
    /// This method is designed for use as a SQL CLR scalar function and can safely be registered in SQL Server
    /// using <c>PERMISSION_SET = SAFE</c>. It may also be used directly from .NET applications.
    /// </para>
    /// <para>
    /// The resulting values are roughly time-ordered, which improves clustered index insert performance compared
    /// to purely random GUIDs. However, ordering across systems assumes reasonably synchronized clocks.
    /// </para>
    /// </remarks>
    /// <returns>
    /// A <see cref="SqlGuid"/> representing a COMB GUID value suitable for use as a SQL Server
    /// <c>uniqueidentifier</c> primary key or default column value.
    /// </returns>
    public static Guid NewGuid()
    {
        // 1) Get current UTC ms and per-ms sequence (thread-safe)
        ulong milliseconds;
        uint localSeq;
        var nowMilliseconds = (ulong)DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

        lock (Lock)
        {
            if ((ulong)_lastMilliseconds == nowMilliseconds)
                _sequence++;
            else
            {
                _lastMilliseconds = (long)nowMilliseconds;
                _sequence = 0;
            }
            milliseconds = nowMilliseconds;
            localSeq = _sequence;
        }

        // 2) Build 8-byte big-endian timestamp
        var timestamp = new byte[8];
        for (int i = 7; i >= 0; i--)
        {
            timestamp[i] = (byte)(milliseconds & 0xFF);
            milliseconds >>= 8;
        }

        // 3) Create 16 random bytes
        Span<byte> bytes = new byte[16];
        RandomNumberGenerator.Fill(bytes);

        // 4) Overlay timestamp bytes into SQL Server compare order
        for (int i = 0; i < 8; i++)
        {
            bytes[SqlOrder[i]] = timestamp[i];
        }

        // 5) Tie-breaker nibble in the last-compared byte (byte 15)
        bytes[15] = (byte)((byte)(bytes[15] & 0xF0) | (byte)(localSeq & 0x0F));

        return new Guid(bytes);
    }
}

File: TaskList.WebApp/src/CombGuid.ts
const SQL_ORDER = [3, 2, 1, 0, 5, 4, 7, 6, 8, 9, 10, 11, 12, 13, 14, 15];

export default class CombGuid {
  private static lastMilliseconds: bigint = 0n;
  private static sequence: number = 0; // 0..15 nibble

  public static newGuid(): string {
    const nowMilliseconds = BigInt(Date.now());
    if (nowMilliseconds === CombGuid.lastMilliseconds) {
      CombGuid.sequence = (CombGuid.sequence + 1) & 0x0f; // 0..15
    } else {
      CombGuid.lastMilliseconds = nowMilliseconds;
      CombGuid.sequence = 0;
    }

    const timestamp = new Uint8Array(8);
    let temporaryMilliseconds = nowMilliseconds;
    for (let i = 7; i >= 0; i--) {
      timestamp[i] = Number(temporaryMilliseconds & 0xffn);
      temporaryMilliseconds >>= 8n;
    }

    const bytes = CombGuid.random16();
    for (let i = 0; i < 8; i++) {
      bytes[SQL_ORDER[i]] = timestamp[i];
    }

    bytes[15] = (bytes[15] & 0xf0) | (CombGuid.sequence & 0x0f);

    const b = bytes;
    return (
      CombGuid.hex2(b[3]) +
      CombGuid.hex2(b[2]) +
      CombGuid.hex2(b[1]) +
      CombGuid.hex2(b[0]) +
      "-" +
      CombGuid.hex2(b[5]) +
      CombGuid.hex2(b[4]) +
      "-" +
      CombGuid.hex2(b[7]) +
      CombGuid.hex2(b[6]) +
      "-" +
      CombGuid.hex2(b[8]) +
      CombGuid.hex2(b[9]) +
      "-" +
      CombGuid.hex2(b[10]) +
      CombGuid.hex2(b[11]) +
      CombGuid.hex2(b[12]) +
      CombGuid.hex2(b[13]) +
      CombGuid.hex2(b[14]) +
      CombGuid.hex2(b[15])
    ).toLowerCase();
  }

  private static hex2(b: number) {
    return b.toString(16).padStart(2, "0");
  }

  private static random16(): Uint8Array {
    const b = new Uint8Array(16);
    crypto.getRandomValues(b);
    return b;
  }
}
